## Introduction

In this article we will learn how to **clean a dataset** i.e. some mandatory steps required that one has to know before getting started with the data preparation phase of data science project in this way the newbies who have started their journey will understand the critical steps that are involved in cleaning/preparing the dataset for further process like **feature engineering**, **machine learning model development**.

In a nuthsell we will try to build a **template** of data cleaning using Python and it's famous libraries (**Numpy and Pandas**).

The very first thing that we should do before actually getting started with data cleaning is - **Ask Questions to yourself** i.e. **What insights you wanna draw from your analysis?**, **what needs to be corrected based on the business requirements?**. Similarly, we have curated down some of the key tasks for us (particularly for this instance):

* How does the dataset handle invalid values? 
* What do we want to do with null values?
* Do we want to summarise, group or filter the data? 

## Importing libraries required for data cleaning

In [1]:
import pandas as pd
import numpy as np

**Pandas** and **Numpy** are the most recommended and powerful library when it comes to dealing with structured data or the **tabular data**

1. **Pandas:** Pandas is tagged as the data manipultion library by every data analyst (both experienced and freshers) as it is enriched with vital functionalities and more than **everything to deal with tabular data**.


2. **Numpy:** Numpy is mainly used for working with arrays and dealing with **complex mathematical calculations** by providing some useful functions linear algebra, fourier transform, and matrices.

In [2]:
df = pd.read_csv("Diabetes.csv")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


**Inference:** Firstly we started with reading the dataset (this is the famous **Diabetes dataset from UCI machine learning repository**) using **read.csv()** function. Then for further intervention using the **info()** method which is giving us following insights:

* Do our dataset have any null values? Answer is no! as we can see **768 non-null values**.
* Real data types are detected for each columns like for **BMI** and **Pedigree function** it is **float** for rest it is **integer**

In [3]:
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

**Inference:** So it looks like they are using **0** values where they dont have data. I don't think theyre using **NaN** at all, but if they were we could either drop those rows (**dropna**) or fill them to some value (**fillna**). Because they're using **0** already, it might be prudent to do this just in case.

In [4]:
df = df.fillna(0)

**Inference:** So now we know there are **NaN** values. We could also have just, you know, checked for **NaN**, but now I'm trying to show functions you can use.

**So, what to do with these zero values?** In some cases we could fill them with something sensible, but that normally just biases the data. So mostly we'd ignore them. So what we want to do is ask **how** we want to use the data. Will we be using **SkinThickness**? Do we care if there are non-physical outliers?

If we cared about Glucose, BMI and Age primarily, we could get rid of a ton of these issues but only looking at those columns.

In [14]:
df2 = df[["Glucose", "BMI", "Age", "Outcome"]]
df2.head()

Glucose   BMI  Age  Outcome
0      148  33.6   50        1
1       85  26.6   31        0
2      183  23.3   32        1
3       89  28.1   21        0
4      137  43.1   33        1

**Inference:** Here we have grabbed the columns in which we are more interested to do our analysis and this can be very common and essential steps in any data cleaning process as we don't need all the columns in hand to deal with. Hence selecting the right set of data is very important.

One can notice that We chose the **Glucose**, **BMI**, **Age**, and **Outcome** column where the last one is target and the rest are features, they are extracted using **multi-indexing**

In [7]:
df2.describe()

Glucose         BMI         Age     Outcome
count  768.000000  768.000000  768.000000  768.000000
mean   120.894531   31.992578   33.240885    0.348958
std     31.972618    7.884160   11.760232    0.476951
min      0.000000    0.000000   21.000000    0.000000
25%     99.000000   27.300000   24.000000    0.000000
50%    117.000000   32.000000   29.000000    0.000000
75%    140.250000   36.600000   41.000000    1.000000
max    199.000000   67.100000   81.000000    1.000000

**Inference:** But now lets get rid of any stray zeros. What we want to do is find a row with **any** number of zeros and remove that row. Or in terms of applying a mask, find the rows which have any **zero (True)**, invert that **(to False)** so when we apply the mask using **loc**, the False entries get dropped.

In [8]:
df3 = df2.loc[~(df2[df2.columns[:-1]] == 0).any(axis=1)]
df3.describe()
df3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 752 entries, 0 to 767
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Glucose  752 non-null    int64  
 1   BMI      752 non-null    float64
 2   Age      752 non-null    int64  
 3   Outcome  752 non-null    int64  
dtypes: float64(1), int64(3)
memory usage: 29.4 KB


**Inference:** Great, so we've selected the data we cared about, made sure it has no **null-like** values. We'll go on to checking things look sane with some plots in the next section. One final thing we could do is to **group the data by outcome**. It might make it easier to look for patterns in diagnoses.

We can do this either by splitting out the DataFrame into two (one for yes and one for no), or if we wanted summary statistics we could use the **groupBy** function:

In [9]:
df3.groupby("Outcome").mean()

Glucose        BMI        Age
Outcome                                  
0        110.825820  30.876434  31.309426
1        142.488636  35.371970  37.015152

**Inference:** And what this can tell us is that, in general, the **higher your glucose level, the more overweight you are**, and the **older you are, the greater your chance of being diagnosed with diabetes**. Which, perhaps, is not that surprising.

We can do other things using the **groupby** statement, like so:

In [10]:
df3.groupby("Outcome").agg({"Glucose": "mean", "BMI": "median", "Age": "sum"})

Glucose    BMI    Age
Outcome                          
0        110.825820  30.10  15279
1        142.488636  34.25   9772

**Inference:** There are multiple ways of using the **GroupBy** function this is the second way, where we can use **agg** function to help us with utilising multiple **central tendency** method like **mean**, **median**, **sum** and so on.

Same thing is visible in the output as well where Glucose, BMI and age are grouped by all three mean, median and sum.

In [11]:
df3.groupby("Outcome").agg(["mean", "median"])

Glucose               BMI               Age       
               mean median       mean median       mean median
Outcome                                                       
0        110.825820  107.5  30.876434  30.10  31.309426     27
1        142.488636  140.5  35.371970  34.25  37.015152     36

**Inference:** We can also see more than one statistical method using **.agg(["method1", "method2",..."methodn")**.

We can also split the dataset into positive and negative outcomes if we wanted. *If*.

In [16]:
positive = df3.loc[df3["Outcome"] == 1]
negative = df3.loc[df3["Outcome"] == 0]
print(positive.shape, negative.shape)

(264, 4) (488, 4)


**Inference:** For knowing that whether our **dataset is balanced or not** we need to see how many **positive outcomes** are there and how many **negative**. From the above output we can conclude that dataset is **imbalanced** as **negative** cases > **positive** cases.

We won't use this splitting just yet, so lets save out the cleaned and prepared dataset, **df3** to file, so our analysis code can load it in the future without having to copy-paste the data prep code into future notebooks.

In [13]:
df3.to_csv("clean_diabetes.csv", index=False)

## Conclusion

Here we are in the last section of this article where we will try to **summarize everything** we did so far so that one would get a rough idea and a sort of revison and end goal that we reached by the closure of this blog. Let's have a look in a nutshell what kind of template we build:

1. First we started off by getting some statistical insights from a dataset by using method like **info()** and **describe()** then we had a look over **head of the dataset** to have a look and feel of the same.


2. Then we move forward to **deal with the missing values** (in our case it was **0**) so first we analyse that and removed those rows which have unsual zeros.


3. At the last we also learned about different ways of grouping the dataset on top of **target** column using all the valid **groupBy function like mean, median or sum** later we saved the cleaned data in the **format of CSV** so that we can use that for further analysis.